### PR TITLE
Added implementation of expiration extension to FileStore

### PR DESCRIPTION
**Pull request by bypie5**
Server owners to configure expiration time with expirationPeriodMintues property in FileStoreOptions. Then, the server owner can periodically invoke Server::cleanUpExpiredUploads() to remove uploads that are both expired AND incomplete.

### CODE REVIEW

**Comment by bypie5** [comment]
Hello, I closed a [similar PR](https://github.com/tus/tus-node-server/pull/315) a few days ago because the `1.x` branch was not yet converted to typescript. Now that it #308 has been merged in, I have reopened this PR with some minor changes.

**Comment by Murderlon** [comment]
Just wanted to say thanks for making this and I'll look into this soon

File: lib/stores/FileStore.ts [hunk 1]
```diff
@@ -32,16 +34,19 @@ const log = debug('tus-node-server:stores:filestore')
...
   directory: string
   configstore: Store
+  expirationPeriodInMinutes: number
 
-  constructor({directory, configstore}: Options) {
+  constructor({directory, configstore, expirationPeriodMinutes}: Options) {
     super()
     this.directory = directory
     this.configstore = configstore ?? new Configstore(`${pkg.name}-${pkg.version}`)
+    this.expirationPeriodInMinutes = expirationPeriodMinutes ?? -1
```
Comment (Murderlon) [inline lib/stores/FileStore.ts:43]: not sure if we want this in minutes instead of miliseconds?

File: lib/stores/FileStore.ts [hunk 2]
```diff
@@ -188,4 +193,42 @@ export default class FileStore extends DataStore {
 
     this.configstore.set(file_id, file)
   }
+
+  async deleteExpired(): Promise<number> {
+    const now = new Date()
+    const toDelete: Promise<void>[] = []
+    for (const file_id of Object.keys(this.configstore.all)) {
+      this.getUpload(file_id)
+        .then((stats: File) => {
```
Comment (Murderlon) [inline lib/stores/FileStore.ts:202]: if we make this an async function, then we shouldn't mix in promise callbacks

File: lib/handlers/PostHandler.ts [hunk 3]
```diff
@@ -68,10 +68,27 @@ export default class PostHandler extends BaseHandler {
...
+        // Value MUST be in RFC 7231 datetime format
+        optional_headers['Upload-Expires'] = new Date(
+          creation.getTime() + this.store.getExpiration() * 60_000
+        ).toUTCString()
+      }
+    }
 
-    // The request MIGHT include a Content-Type header when using creation-with-upload extension
-    if (!RequestValidator.isInvalidHeader('content-type', req.headers['content-type'])) {
+    if (RequestValidator.isInvalidHeader('content-type', req.headers['content-type'])) {
```
Comment (Murderlon) [inline lib/handlers/PostHandler.ts:87]: I would leave this as it was, also with the deleted comment, and start a new block specifically for expiry. I find this flow control hard to follow.

**Review by Murderlon — COMMENTED** [review]
Wouldn't it make more sense to store the expiration instead of the creation date in the file model? Then we don't do `creation.getTime() + this.getExpiration() * 60_000` every time and we can remove the `getExperiration` method?

**Comment by Murderlon** [comment]
We would also need a couple of tests for this

File: lib/stores/FileStore.ts (hunk 1 above)
Comment (bypie5) [reply lib/stores/FileStore.ts]: I changed the option to represent milliseconds instead of minutes. 

Also, the reason I am computing `const expires = new Date(creation.getTime() + this.getExpiration())` with the `creation` time instead of storing expiration date is because the expiration period can change over time, "If the expiration is known at the creation, the Upload-Expires header MUST be included in the response to the initial POST request. Its value MAY change over time."

I figured it would be easier to compute the expiration time on the fly with the most current value for `expirationPeriodInMilliseconds` than to save an outdated date in the configstore.

**Comment by bypie5** [comment]
Some End-to-end tests have been added to test this feature

**Comment by mitjap** [comment]
Should `PATCH`, `HEAD` and possibly other request check this expiration date and return 404 even if the file might not be removed at the time? I can also think of various race conditions that can happen when cleanup occurs while user is still uploading a file. 

Should active uploads be aborted when expiration time is reached?

Try this (extreme and unlikely) scenario. Run `deleteExpired` every second or so. Set expiration to one second. Upload file that will take at least 10 second, possibly in multiple chunks. :thinking: Expected behavior is that upload will be removed in between chunks and next chunk will return error 404. But in reality I guess there will (probably) be some kind of server error.

This can cause weird and hard to track errors. Especially with other stores that might follow similar pattern when implementing.

**Comment by mitjap** [comment]
Is it possible to make expiration extension datastore agnostic? I think every datastore can implement `expiration` extension (can share same `deleteExpired` code) given that it implements `termination` extension (has `remove` method`) and has enumerable configstore.

File: lib/stores/FileStore.ts [hunk 4]
```diff
@@ -17,11 +17,13 @@ type Store = {
   get(key: string): File | undefined
   set(key: string, value: File): void
   delete(key: string): void
+  all: Record<string, File>
```
Comment (mitjap) [inline lib/stores/FileStore.ts:20]: I think this should not return all records since this would require that all records are read to memory. Maybe `all: string[]` would be more suitable? This would return only the keys.

File: lib/stores/FileStore.ts [hunk 5]
```diff
@@ -188,4 +193,48 @@ export default class FileStore extends DataStore {
...
+          'creation_date' in info &&
+          this.getExpiration() > 0 &&
+          stats.size !== info.upload_length &&
+          stats.id === info.id
+        ) {
+          const creation = new Date(info.creation_date as Date)
+          const expires = new Date(creation.getTime() + this.getExpiration())
+          if (now > expires) {
+            toDelete.push(this.remove(file_id))
+            this.configstore.delete(file_id)
```
Comment (mitjap) [inline lib/stores/FileStore.ts:222]: I don't think it is safe to remove an upload from configstore while `remove`'s promise is still pending.  Current implementation of `remove` does not check configstore but I would not rely on this behavior.

When sucsessfully completed `remove` will remove file from configstore.

File: lib/stores/FileStore.ts [hunk 6]
```diff
@@ -188,4 +193,48 @@ export default class FileStore extends DataStore {
...
+      Object.keys(this.configstore.all).map((file_id) => {
+        return this.getUpload(file_id)
+      })
+    )
+
+    let index = 0
+    for (const file_id of Object.keys(this.configstore.all)) {
+      try {
+        const stats = fileStats[index]
+        const info = this.configstore.get(file_id)
```
Comment (mitjap) [inline lib/stores/FileStore.ts:210]: You call `this.configstore.all` two times and every time it returns all data which you discard and then later call `this.configstore.get` to retrieve discarded data. Note that everytime you read from or write to configstore (https://www.npmjs.com/package/configstore) entire file is read and parsed from disk or serialized and written to disk.

Given that N files are stored in configstore, this function does 2 + N (sync) reads and 2* N (sync) writes. I think this can be optimized to 1 read (if you leave `this.configstore.all` as is - see other comment) and N writes.

File: lib/stores/FileStore.ts [hunk 7]
```diff
@@ -188,4 +193,48 @@ export default class FileStore extends DataStore {
...
+    const now = new Date()
+    const toDelete: Promise<void>[] = []
+    const fileStats = await Promise.all(
+      Object.keys(this.configstore.all).map((file_id) => {
+        return this.getUpload(file_id)
+      })
+    )
+
+    let index = 0
+    for (const file_id of Object.keys(this.configstore.all)) {
```
Comment (mitjap) [inline lib/stores/FileStore.ts:207]: I would recommend against calling this.configstore.all multiple times as content can and will change between calls.
When content of configstore changes, indices will not match anymore and entire loop will fail to do anything meaningful.

File: lib/models/File.ts [hunk 8]
```diff
@@ -4,6 +4,7 @@ export default class File {
   upload_metadata?: string
   upload_length?: string
   size?: number
+  creation_date?: Date
```
Comment (mitjap) [inline lib/models/File.ts:7]: Is this really optional? This property is set in constructor every time.

File: lib/stores/FileStore.ts [hunk 9]
```diff
@@ -32,16 +34,19 @@ const log = debug('tus-node-server:stores:filestore')
...
   directory: string
   configstore: Store
+  expirationPeriodInMilliseconds: number
 
-  constructor({directory, configstore}: Options) {
+  constructor({directory, configstore, expirationPeriodInMilliseconds}: Options) {
     super()
     this.directory = directory
     this.configstore = configstore ?? new Configstore(`${pkg.name}-${pkg.version}`)
+    this.expirationPeriodInMilliseconds = expirationPeriodInMilliseconds ?? -1
```
Comment (mitjap) [inline lib/stores/FileStore.ts:43]: Default value in `DataStore` is `0`. Why not keep same default values?

**Comment by bypie5** [comment]
@mitjap To address your point:

> Should PATCH, HEAD and possibly other requests check this expiration date and return 404 even if the cleanup function has not (yet) been called? 

I think we might want to return 410 in those cases instead of 404. My reasoning is because of the following language in the specs:

> If a Client does attempt to resume an upload which has since been removed by the Server, the Server SHOULD respond with the404 Not Found or 410 Gone status. The latter one SHOULD be used if the Server is keeping track of expired uploads. In both cases the Client SHOULD start a new upload. - Tus specs

But that is a good point you brought up about cleaning up an upload while someone is still trying to send data. 

> Should active uploads be aborted when expiration time is reached?

I think it would make sense to have expired uploads be aborted and report some error (410) back to the client. I will go ahead and try your example too and report back some observations.

**Comment by Acconut** [comment]
Wow, that is an amazing PR! As a general question, we should decide if the upload expiration should be counted since the creation time or the last modification time. The spec allows both and this PR implements the former. Just to double check that this is the way we want to implement it right now.

File: lib/stores/FileStore.ts (hunk 9 above)
Comment (bypie5) [reply lib/stores/FileStore.ts]: Both values are now set to `0`.

File: lib/models/File.ts (hunk 8 above)
Comment (bypie5) [reply lib/models/File.ts:7]: Good point - I don't think this property should be optional.

Part of the reason I made `creation_date` optional was because there were a few places in `GCSDataStore` and `S3Store` that were initializing `File` objects with bracket notations:

Screen Shot 2022-11-05 at 5 00 12 PM

Making the creation_date required means that those places need to be defined with an appropriate value. At the time of creating this PR, I didn't want to make modifications outside of the `FileStore` file. However, since [your comment above](https://github.com/tus/tus-node-server/pull/320#issuecomment-1301192982) suggests looking into making the expiration code DataStore agnostic, I will also look into fixing those definitions of `File` objects when creation_date is marked as required.

**Comment by Murderlon** [comment]
Thanks for applying the feedback and writing tests. This is starting to look really good. There are currently some merge conflicts because I renamed the `File` model to `Upload` and changed the names in there. The types should be a lot better but unfortunately there are now some conflicts.

File: lib/stores/FileStore.ts (hunk 6 above)
Comment (bypie5) [reply lib/stores/FileStore.ts]: Thanks for pointing out that the whole file is being read on configstore accesses! I refactored that portion of the code to only read from config store once like you suggested.

After that, I did some timing of `HEAD` requests of a server with 1k and 10k files. I timed the requests before the files were deleted, while the files were being deleted, and after all files had been deleted. Here are the summarized results.

In both cases you can see a small differences in time between `HEAD` requests made with a non-zero number of uploads saved in the configstore. Additionally there is a much larger difference in time between `HEAD` requests while the server is deleting and when it is not.

**1000 files: max stall time 845ms || max stall time before deleting 33ms || max stall time after deleting 10ms**
```
(now: Sun Nov 13 2022 16:17:15 GMT-0800 (Pacific Standard Time)) head request took [18ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:15 GMT-0800 (Pacific Standard Time)) head request took [17ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:15 GMT-0800 (Pacific Standard Time)) head request took [18ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:17 GMT-0800 (Pacific Standard Time)) head request took [16ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:17 GMT-0800 (Pacific Standard Time)) head request took [15ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:24 GMT-0800 (Pacific Standard Time)) head request took [19ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:24 GMT-0800 (Pacific Standard Time)) head request took [18ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:24 GMT-0800 (Pacific Standard Time)) head request took [19ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:26 GMT-0800 (Pacific Standard Time)) head request took [31ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:26 GMT-0800 (Pacific Standard Time)) head request took [32ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:26 GMT-0800 (Pacific Standard Time)) head request took [33ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:27 GMT-0800 (Pacific Standard Time)) head request took [20ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:27 GMT-0800 (Pacific Standard Time)) head request took [19ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:27 GMT-0800 (Pacific Standard Time)) head request took [18ms] size on disk: 1
// SERVER HAS ABOUT 1k uploads

// STARTED DELETING 1000 expired uploads
(now: Sun Nov 13 2022 16:17:29 GMT-0800 (Pacific Standard Time)) head request took [523ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:29 GMT-0800 (Pacific Standard Time)) head request took [522ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:29 GMT-0800 (Pacific Standard Time)) head request took [523ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:29 GMT-0800 (Pacific Standard Time)) head request took [522ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:31 GMT-0800 (Pacific Standard Time)) head request took [1255ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:31 GMT-0800 (Pacific Standard Time)) head request took [1254ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:31 GMT-0800 (Pacific Standard Time)) head request took [1253ms] size on disk: 1
(now: Sun Nov 13 2022 16:17:31 GMT-0800 (Pacific Standard Time)) head request took [1254ms] size on disk: 1
HEAD http://localhost:1080/files/73eac9cbb34523adcb603336870ad53e net::ERR_ABORTED 410 (Gone)
(now: Sun Nov 13 2022 16:17:32 GMT-0800 (Pacific Standard Time)) head request took [908ms] size on disk: null
HEAD http://localhost:1080/files/a9b1963cad5c1d6801d19b8bfe595730 net::ERR_ABORTED 410 (Gone)
(now: Sun Nov 13 2022 16:17:32 GMT-0800 (Pacific Standard Time)) head request took [912ms] size on disk: null
HEAD http://localhost:1080/files/3e28dfe79bef43cd16fa70b25415790f net::ERR_ABORTED 410 (Gone)
(now: Sun Nov 13 2022 16:17:32 GMT-0800 (Pacific Standard Time)) head request took [912ms] size on disk: null
HEAD http://localhost:1080/files/2b47e539fd2d0957e1f607e897a8d814 net::ERR_ABORTED 410 (Gone)
(now: Sun Nov 13 2022 16:17:32 GMT-0800 (Pacific Standard Time)) head request took [912ms] size on disk: null
HEAD http://localhost:1080/files/3b939a30c15e99c7e35c2d7e9e27599e net::ERR_ABORTED 410 (Gone)
(now: Sun Nov 13 2022 16:17:32 GMT-0800 (Pacific Standard Time)) head request took [912ms] size on disk: null
HEAD http://localhost:1080/files/fe424f6f0859b2ab31d07f92d6ef2339 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:33 GMT-0800 (Pacific Standard Time)) head request took [280ms] size on disk: null
HEAD http://localhost:1080/files/27f8cab99b364753e7e1580f07f6893f net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:33 GMT-0800 (Pacific Standard Time)) head request took [290ms] size on disk: null
HEAD http://localhost:1080/files/33acc6cdad8e7dafb36efadf3b7b2011 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:33 GMT-0800 (Pacific Standard Time)) head request took [293ms] size on disk: null
HEAD http://localhost:1080/files/29dc9c4361e569f167a654bcb512ef55 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:33 GMT-0800 (Pacific Standard Time)) head request took [295ms] size on disk: null
HEAD http://localhost:1080/files/5579eec897d94550ddff092c61efc170 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:33 GMT-0800 (Pacific Standard Time)) head request took [296ms] size on disk: null
HEAD http://localhost:1080/files/e21dcb5f543308f3bcf515698b6944e7 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:35 GMT-0800 (Pacific Standard Time)) head request took [755ms] size on disk: null
HEAD http://localhost:1080/files/e3468d2ef2be002dfecdfe1f41f2c67a net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:35 GMT-0800 (Pacific Standard Time)) head request took [754ms] size on disk: null
HEAD http://localhost:1080/files/5df07d1acc516bee909b7bb084455250 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:35 GMT-0800 (Pacific Standard Time)) head request took [755ms] size on disk: null
HEAD http://localhost:1080/files/d9e8495da7f037be255a885164686f09 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:35 GMT-0800 (Pacific Standard Time)) head request took [756ms] size on disk: null
HEAD http://localhost:1080/files/3d0a16f0d9ded9811852b910f4f81a95 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:35 GMT-0800 (Pacific Standard Time)) head request took [757ms] size on disk: null
HEAD http://localhost:1080/files/2a8c3675b6b8ec6ad116aad30f0ca4aa 404 (Not Found)
HEAD http://localhost:1080/files/f8ef5564d5f482713d34ba6cee9f614f net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:38 GMT-0800 (Pacific Standard Time)) head request took [842ms] size on disk: null
HEAD http://localhost:1080/files/1c77fce6374dce8f2f89be1ed83086da net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:38 GMT-0800 (Pacific Standard Time)) head request took [842ms] size on disk: null
HEAD http://localhost:1080/files/adb586616bbc3e24b78fa4372ee1a217 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:38 GMT-0800 (Pacific Standard Time)) head request took [844ms] size on disk: null
HEAD http://localhost:1080/files/c791175fd34635b3ab99396a23fed31f net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:38 GMT-0800 (Pacific Standard Time)) head request took [845ms] size on disk: null
HEAD http://localhost:1080/files/29100a97f095419c284d4ed3ef43aacc net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:38 GMT-0800 (Pacific Standard Time)) head request took [845ms] size on disk: null
HEAD http://localhost:1080/files/f0e86f2684504588a4733e21ddd0cf65 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:39 GMT-0800 (Pacific Standard Time)) head request took [296ms] size on disk: null
HEAD http://localhost:1080/files/8c5ddb5a5de4c387a7fc4a941af7ffb2 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:39 GMT-0800 (Pacific Standard Time)) head request took [296ms] size on disk: null
HEAD http://localhost:1080/files/ad869753bd81c9d66ce267aaed7f3fcc net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:39 GMT-0800 (Pacific Standard Time)) head request took [296ms] size on disk: null
HEAD http://localhost:1080/files/74d29c593573647f45c6f526a97553b0 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:39 GMT-0800 (Pacific Standard Time)) head request took [297ms] size on disk: null
HEAD http://localhost:1080/files/073b13c262cf04dd8ec4e5acc6764e45 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:39 GMT-0800 (Pacific Standard Time)) head request took [297ms] size on disk: null
HEAD http://localhost:1080/files/8d7af3c2249aba9fea52e3c833be349c 404 (Not Found)
(now: Sun Nov 13 2022 16:17:42 GMT-0800 (Pacific Standard Time)) head request took [75ms] size on disk: null
HEAD http://localhost:1080/files/e4ab5298539a2791d85514b42fbb4c92 404 (Not Found)
(now: Sun Nov 13 2022 16:17:44 GMT-0800 (Pacific Standard Time)) head request took [534ms] size on disk: null
HEAD http://localhost:1080/files/8ddf3eb41f998bd45f63122929ba78ce net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:44 GMT-0800 (Pacific Standard Time)) head request took [534ms] size on disk: null
HEAD http://localhost:1080/files/472a804663ac224f044282d6189924d4 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:44 GMT-0800 (Pacific Standard Time)) head request took [535ms] size on disk: null
HEAD http://localhost:1080/files/18053a1d1fd92426cc83631736d1fb37 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:45 GMT-0800 (Pacific Standard Time)) head request took [344ms] size on disk: null
HEAD http://localhost:1080/files/92039f3f680ac9799193702f86e744df 404 (Not Found)
HEAD http://localhost:1080/files/eeff787ba3f7d14f808679fef6d451df net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:48 GMT-0800 (Pacific Standard Time)) head request took [191ms] size on disk: null
(now: Sun Nov 13 2022 16:17:50 GMT-0800 (Pacific Standard Time)) head request took [197ms] size on disk: null
HEAD http://localhost:1080/files/309dac4c73a7082ccb0a6a78b99ac7c8 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:50 GMT-0800 (Pacific Standard Time)) head request took [195ms] size on disk: null
HEAD http://localhost:1080/files/dcdae6b739f7dcf1d0ff9fd9273f7f5e net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:50 GMT-0800 (Pacific Standard Time)) head request took [195ms] size on disk: null
HEAD http://localhost:1080/files/82f1c4e8124943821d3f17bb15c6c3e7 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:50 GMT-0800 (Pacific Standard Time)) head request took [195ms] size on disk: null
HEAD http://localhost:1080/files/d48111a4ea88341a666bfd50bf83c2e5 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:50 GMT-0800 (Pacific Standard Time)) head request took [194ms] size on disk: null
HEAD http://localhost:1080/files/471b01c7187512b0d08386f268b232b3 net::ERR_ABORTED 404 (Not Found)
// DONE DELETING EXPIRED FILES
(now: Sun Nov 13 2022 16:17:51 GMT-0800 (Pacific Standard Time)) head request took [10ms] size on disk: null
HEAD http://localhost:1080/files/30fc791a606842f9e68637e53767e6d6 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:51 GMT-0800 (Pacific Standard Time)) head request took [9ms] size on disk: null
HEAD http://localhost:1080/files/40926a51be87a5b42c14d491f1ace424 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:51 GMT-0800 (Pacific Standard Time)) head request took [9ms] size on disk: null
HEAD http://localhost:1080/files/f341e54b43e0ef0f0585ad514f54d442 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:51 GMT-0800 (Pacific Standard Time)) head request took [10ms] size on disk: null
HEAD http://localhost:1080/files/1d2efc73772ff855a87853a3d1f9d97a net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:51 GMT-0800 (Pacific Standard Time)) head request took [9ms] size on disk: null
HEAD http://localhost:1080/files/7bdfd9b697ea6308b2dd2c91d8332e3b net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:52 GMT-0800 (Pacific Standard Time)) head request took [9ms] size on disk: null
HEAD http://localhost:1080/files/1881b009ef577f01daf4885d6c1d5ee6 net::ERR_ABORTED 404 (Not Found)
(now: Sun Nov 13 2022 16:17:52 GMT-0800 (Pacific Standard Time)) head request took [8ms] size on disk: null
HEAD http://localhost:1080/files/bf8b3b2503e005e56282ff6a1b0807fc net::ERR_ABORTED 404 (Not Found)
```

**10000 files: max stall time 93492ms || max stall time before deleting 73ms || max stall time after deleting 13ms**
```
[Sun Nov 13 2022 17:51:42 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads
[Sun Nov 13 2022 17:51:52 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads
[Sun Nov 13 2022 17:52:02 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads
[Sun Nov 13 2022 17:52:12 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads
[Sun Nov 13 2022 17:52:22 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads
[Sun Nov 13 2022 17:52:32 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads -> (now: Sun Nov 13 2022 17:52:38 GMT-0800 (Pacific Standard Time)) head request took [73ms]
[Sun Nov 13 2022 17:52:42 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads -> (now: Sun Nov 13 2022 17:52:43 GMT-0800 (Pacific Standard Time)) head request took [68ms]
[Sun Nov 13 2022 17:52:52 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads
// CONFIGSTORE HAS ~10_000 entries

// SERVER STARTS DELETING EXPIRED UPLOADS EVERY 10 SECONDS
[Sun Nov 13 2022 17:53:12 GMT-0800 (Pacific Standard Time)] Deleted 384 expired uploads
[Sun Nov 13 2022 17:53:23 GMT-0800 (Pacific Standard Time)] Deleted 428 expired uploads
[Sun Nov 13 2022 17:53:35 GMT-0800 (Pacific Standard Time)] Deleted 458 expired uploads -> (now: Sun Nov 13 2022 17:53:35 GMT-0800 (Pacific Standard Time)) head request took [28089ms]
[Sun Nov 13 2022 17:53:48 GMT-0800 (Pacific Standard Time)] Deleted 522 expired uploads
[Sun Nov 13 2022 17:54:02 GMT-0800 (Pacific Standard Time)] Deleted 601 expired uploads
[Sun Nov 13 2022 17:54:18 GMT-0800 (Pacific Standard Time)] Deleted 662 expired uploads -> (now: Sun Nov 13 2022 17:54:18 GMT-0800 (Pacific Standard Time)) head request took [60740ms]
[Sun Nov 13 2022 17:54:33 GMT-0800 (Pacific Standard Time)] Deleted 661 expired uploads
[Sun Nov 13 2022 17:54:46 GMT-0800 (Pacific Standard Time)] Deleted 601 expired uploads
[Sun Nov 13 2022 17:54:57 GMT-0800 (Pacific Standard Time)] Deleted 542 expired uploads
[Sun Nov 13 2022 17:55:06 GMT-0800 (Pacific Standard Time)] Deleted 432 expired uploads -> (now: Sun Nov 13 2022 17:55:01 GMT-0800 (Pacific Standard Time)) head request took [93492ms]
[Sun Nov 13 2022 17:55:16 GMT-0800 (Pacific Standard Time)] Deleted 361 expired uploads
[Sun Nov 13 2022 17:55:24 GMT-0800 (Pacific Standard Time)] Deleted 336 expired uploads
[Sun Nov 13 2022 17:55:33 GMT-0800 (Pacific Standard Time)] Deleted 317 expired uploads
[Sun Nov 13 2022 17:55:43 GMT-0800 (Pacific Standard Time)] Deleted 311 expired uploads
[Sun Nov 13 2022 17:55:54 GMT-0800 (Pacific Standard Time)] Deleted 308 expired uploads
[Sun Nov 13 2022 17:56:04 GMT-0800 (Pacific Standard Time)] Deleted 293 expired uploads -> (now: Sun Nov 13 2022 17:56:04 GMT-0800 (Pacific Standard Time)) head request took [6050ms]
[Sun Nov 13 2022 17:56:14 GMT-0800 (Pacific Standard Time)] Deleted 293 expired uploads
[Sun Nov 13 2022 17:56:23 GMT-0800 (Pacific Standard Time)] Deleted 287 expired uploads
[Sun Nov 13 2022 17:56:32 GMT-0800 (Pacific Standard Time)] Deleted 281 expired uploads
[Sun Nov 13 2022 17:56:43 GMT-0800 (Pacific Standard Time)] Deleted 278 expired uploads
[Sun Nov 13 2022 17:56:52 GMT-0800 (Pacific Standard Time)] Deleted 274 expired uploads
[Sun Nov 13 2022 17:57:02 GMT-0800 (Pacific Standard Time)] Deleted 261 expired uploads -> (now: Sun Nov 13 2022 17:57:02 GMT-0800 (Pacific Standard Time)) head request took [1483ms]
[Sun Nov 13 2022 17:57:12 GMT-0800 (Pacific Standard Time)] Deleted 261 expired uploads
[Sun Nov 13 2022 17:57:21 GMT-0800 (Pacific Standard Time)] Deleted 260 expired uploads
[Sun Nov 13 2022 17:57:32 GMT-0800 (Pacific Standard Time)] Deleted 262 expired uploads -> (now: Sun Nov 13 2022 17:57:37 GMT-0800 (Pacific Standard Time)) head request took [4ms]
[Sun Nov 13 2022 17:57:41 GMT-0800 (Pacific Standard Time)] Deleted 251 expired uploads
[Sun Nov 13 2022 17:57:48 GMT-0800 (Pacific Standard Time)] Deleted 75 expired uploads
// SERVER DONE DELETING

// CONFIGSTORE HAS 0 ENTRIES
[Sun Nov 13 2022 17:57:57 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads
[Sun Nov 13 2022 17:58:07 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads
[Sun Nov 13 2022 17:58:17 GMT-0800 (Pacific Standard Time)] Deleted 0 expired uploads
```

Raw text data from debug script console output:
[1k.txt](https://github.com/tus/tus-node-server/files/9999011/1k.txt)
[10k.txt](https://github.com/tus/tus-node-server/files/9999014/10k.txt)

File: lib/models/Upload.ts [hunk 10]
```diff
@@ -3,13 +3,15 @@ type TUpload = {
...
   metadata?: string
+  creation_date?: string
 }
 
 export default class Upload {
   id: TUpload['id']
   metadata?: TUpload['metadata']
   size?: TUpload['size']
   offset: TUpload['offset']
+  creation_date?: TUpload['creation_date']
```
Comment (Murderlon) [inline lib/models/Upload.ts:14]: It's never undefined
```suggestion
  creation_date: TUpload['creation_date']
```

File: lib/models/Upload.ts [hunk 11]
```diff
@@ -20,6 +22,8 @@ export default class Upload {
     this.size = upload.size
     this.offset = upload.offset
     this.metadata = upload.metadata
+
+    this.creation_date = upload?.creation_date ?? new Date().toISOString()
```
Comment (Murderlon) [inline lib/models/Upload.ts:26]:
```suggestion
    this.creation_date = upload.creation_date ?? new Date().toISOString()
```

File: lib/stores/FileStore.ts (hunk 4 above)
Comment (mitjap) [reply lib/stores/FileStore.ts]: That's true.

 It is good to have common interface for all possible config stores (could be file, redis or memory based, or really anything else) and for that reason it is not ideal to design such interface based on specific implementation that has very poor performance.

If we go with more general interface for cofig stores then `configstore` could use an adapter to make it compatible.

Comment (Murderlon) [reply lib/stores/FileStore.ts]: Yes. Can do/talk about that in another PR.

**Comment by Murderlon** [comment]
Bit confusing why the tests fail on Node 19.x but not on the others 🤔

**Comment by bypie5** [comment]
> Bit confusing why the tests fail on Node 19.x but not on the others 🤔

I am also confused as well... at least for the failing `PATCH` test requests, it looks they are all erroneously detecting `req.headers['upload-offset']` as undefined when running the test with node 19:
Screen Shot 2022-11-16 at 9 01 02 PM

**Comment by Murderlon** [comment]
I ran the test locally on Node.js 19.0.1 and it ran successful, then upgraded to 19.1.0 and it failed...

They can't make a breaking change like this in a feature version I'd reckon so perhaps it's a Node bug?

**Comment by Murderlon** [comment]
Pinned the CI version to `19.0.1` for now.

**Review by Murderlon — APPROVED** [review]

**Comment by Murderlon** [comment]
I'll merge it tomorrow if no other reviews come in by then :)

**Comment by Murderlon** [comment]
Thanks a lot for this big contribution!
